Loops: Stack-Safe Recursion

fwbrasil opened the pull request:
Kyo's execution model evaluates pure code strictly as [described in the readme](https://github.com/getkyo/kyo/blob/457d9f5958afb6af8b847245164d85de3362e439/README.md?plain=1#L1741), which makes recursive computations without effect suspensions not stack safe by default. The current recommendation is introducing an effect suspension like `IOs` to achieve stack safety.

Although introducing an effect suspension can be convenient in some cases, it can become problematic in code that doesn't have a reasonable way to introduce suspensions. For example, I've been working on the `Streams` effect and adding `IOs` to its execution just to ensure stack safety doesn't seem reasonable. I could emit an empty chunk but that would introduce unnecessary overhead and doesn't feel like a proper solution.

This PR introduces `Loops`, which provides generic stack-safe recursion using a technique identified as part of https://github.com/getkyo/kyo/pull/253 that is currently [used in Kyo's core](https://github.com/getkyo/kyo/blob/457d9f5958afb6af8b847245164d85de3362e439/kyo-core/shared/src/main/scala/kyo/core.scala#L72). The technique reflects on the result of the computation to determine if it can continue the loop immediately (using @tailrec) or if the loop should suspend (using `map`).

hearnadam commented on kyo-core/shared/src/main/scala/kyo/loops.scala:
```diff
@@ -0,0 +1,104 @@
+        loop(input)
+    end transform
+
+    inline def transform2[Input1, Input2, Output: Flat, S](
```
Personally, I dislike `functionN` naming - is there a reason we can't overload the `transform` name?

hearnadam commented on kyo-core/shared/src/test/scala/kyoTest/loopsTest.scala:
```diff
@@ -0,0 +1,293 @@
+
+import kyo.*
+
+class loopsTest extends KyoTest:
```
Perhaps it's worth adding a test with a `List` accumulator - assuming we want users referencing tests. It seems more inline with how I might use this in practical application.

hearnadam commented on README.md:
```diff
@@ -655,6 +655,79 @@ val d: Seq[Int] < Any =
+val d: Int < IOs = 
+  Loops.transform(1)(i => 
+    if i < 5 then 
+      IOs(println(s"Iteration: $i")).map(_ => Loops.continue(i + 1))
```
Can we use `Consoles` here (or in a test)? It may be valuable to highlight that effects other than `IOs` can be combined with Loops. This may be non-obvious when not referencing the code...

Though to be honest, most newcomers may not realize Consoles is a different effect type.

fwbrasil commented:
Thanks for the review! I've pushed changes addressing the points you raised. I've also added `indexed` methods since it's a common case and it can avoid boxing to `Integer`.
